samza-core/src/main/java/org/apache/samza/runtime/LocalContainerRunner.java

```diff
@@ -0,0 +1,122 @@
+ *
+ * Since we don't have the {@link org.apache.samza.coordinator.JobCoordinator} implementation in Yarn, the components (jobModel and containerId)
+ * are directly inside the runner.
+ */
+public class LocalContainerRunner extends AbstractApplicationRunner {
+  private static final Logger log = LoggerFactory.getLogger(LocalContainerRunner.class);
+  private final JobModel jobModel;
+  private final Integer containerId;
+
+  public LocalContainerRunner(JobModel jobModel, Integer containerId) {
+    super(jobModel.getConfig());
+    this.jobModel = jobModel;
+    this.containerId = containerId;
+  }
+
+  @Override
+  public void run(StreamApplication streamApp) {
+    JmxServer jmxServer = null;
+    try {
+      jmxServer = new JmxServer();
+      ContainerModel containerModel = jobModel.getContainers().get(containerId);
+      Object taskFactory = TaskFactoryUtil.createTaskFactory(config, streamApp, this);
+
+      SamzaContainer container = SamzaContainer$.MODULE$.apply(
+          containerModel.getContainerId(),
+          containerModel,
+          config,
+          jobModel.maxChangeLogStreamPartitions,
+          SamzaContainer.getLocalityManager(containerId, config),
+          jmxServer,
+          Util.javaMapAsScalaMap(new HashMap<String, MetricsReporter>()),
+          taskFactory,
+          this);
+
+      container.run();
+    } finally {
+      if (jmxServer != null) {
+        jmxServer.stop();
+      }
+    }
+  }
+
+
+  public static void main(String[] args) throws Exception {
+    setExceptionHandler(() -> System.exit(1));
+
+    Integer containerId = Integer.valueOf(System.getenv(ShellCommandConfig.ENV_CONTAINER_ID()));
+    log.info(String.format("Got container ID: %d", containerId));
+    String coordinatorUrl = System.getenv(ShellCommandConfig.ENV_COORDINATOR_URL());
+    log.info(String.format("Got coordinator URL: %s", coordinatorUrl));
+    int delay = new Random().nextInt(SamzaContainer.DEFAULT_READ_JOBMODEL_DELAY_MS()) + 1;
+    JobModel jobModel = SamzaContainer.readJobModel(coordinatorUrl, delay);
+    Config config = jobModel.getConfig();
+    JobConfig jobConfig = new JobConfig(config);
+    if (jobConfig.getName().isEmpty()) {
+      throw new SamzaException("can not find the job name");
+    }
+    String jobName = jobConfig.getName().get();
+    String jobId = jobConfig.getJobId().getOrElse(ScalaToJavaUtils.defaultValue("1"));
+    MDC.put("containerName", "samza-container-" + containerId);
+    MDC.put("jobName", jobName);
+    MDC.put("jobId", jobId);
+
+    StreamApplication streamApp = TaskFactoryUtil.createStreamApplication(config);
+    new LocalContainerRunner(jobModel, containerId).run(streamApp);
+  }
+
+  /* package private */ static void setExceptionHandler(Runnable runnable) {
+    Thread.UncaughtExceptionHandler exceptionHandler = (t, e) -> {
+      log.error(String.format("Uncaught exception in thread (name=%s). Exiting process now.", t.getName(), e));
+      e.printStackTrace(System.err);
```

Comment:
Why both log error and print stacktrace? Also, I think e.printStackTrace will print to System.err by default.

Maybe use a `Callable` instead of Runnable that's called with the actual exception so that it can act on it?